### ISSUE TITLE
[Discussion] Implementing Async001: AvoidAsyncVoid with IOperation APIs

### ISSUE THREAD

**Issue by jamesqo** [dotnet/roslyn-analyzers#1483]
#### Analyzer package

Microsoft.CodeQuality.Analyzers

#### Analyzer

AvoidAsyncVoid

#### Description

I would like to take on the task (no pun intended) of implementing `AvoidAsyncVoid` using IOperation. The original implementation chooses not to fire when it detects a method with the signature `async void M(object sender, EventArgs e)`, which is reasonable since that typically indicates an event handler hooked up to some UI control.

However, I think there are other cases where `async void` should be considered valid. For example, when writing async xUnit tests, one can return either `void` or `Task` because xUnit uses a custom sync context that allows for both. Do you guys have any ideas for a reasonable way to accommodate for all of these other valid use cases?

/cc @333fred, @mavasani

**Comment by jamesqo** [comment]
My idea: Maybe we should flag an `async void` method only when we can actually find a location it's actually invoked in source. e.g. We'd flag for the following scenarios:

```cs
public async void Callee() {
}

public async Task Caller1() {
    Callee();
}

public void Caller2() {
    Callee();
}
```

We wouldn't flag for these scenarios:

```cs
// We can't find a location in this assembly where Test is actually called, so don't flag it.
[Fact] public async void Test() { }

// Although Handler is referenced through a delegate, don't flag it since it isn't invoked directly.
async void Handler(/* may not follow convention of having 'object sender, EventArgs e' signature */) {}

// Somewhere in source...
_button.Click += Handler;
```

**Disadvantages:** a library could expose an `async void` method that it does not call in its own code, and a method in a different assembly could call it without the analyzer firing.

**Comment by 333fred** [comment]
I think that disadvantage is a pretty strong reason to flag in all scenarios. We've talked before about ignoring certain diagnostics in test methods before. However, I'm not sure this is a place for them. xUnit tests, at least, should be `async Task`, not `async void`, or you risk race conditions.

**Comment by mavasani** [comment]
Async usage analyzers are no longer supported in this repo.
